File: 2015/03/main.js
const path = require("path")
const fs = require("fs")

/**
    Example values

    > = 2 -> initial house + 1
    ^>v< = 4 -> initial + 3 (initial receives two times)
    ^v^v^v^v^v = 2 -> initial + 1 (initial and above receives multiples times)
*/

const filename = "input"
const file = 
    // "^v".trim().split("")
    fs.readFileSync(path.resolve(__dirname, filename), { encoding: "utf-8" }).trim().split("")

const mapped_values = file.map(char => {
    switch (char) {
        case "^":
            return [1, 0]
        case ">":
            return [0, 1]
        case "<":
            return [0, -1]
        case "v":
            return [-1, 0]
        default:
            break
    }
})

function part_one() {
    const visited = {'0,0': 1}
    const position = [0,0]
    for (const house of mapped_values) {
        const [y, x] = house
        position[0] += y
        position[1] += x
        if (visited[position]) {
            visited[position] = visited[position] + 1
            continue
        }
        visited[position] = 1
        
    }
    console.log({ visited: Object.keys(visited).length })
}

function part_two() {
    const visited_santa = {'0,0': 2}
    const position_santa = [0,0]

    const visited_robo = {}
    const position_robo = [0,0]

    for (let i = 0; i < mapped_values.length; i++) {
        const [y, x] = mapped_values[i]
        if (i % 2 === 0) {
            position_santa[0] += y
            position_santa[1] += x
            if (visited_santa[position_santa]) {
                visited_santa[position_santa] = visited_santa[position_santa] + 1
                continue
            }
            visited_santa[position_santa] = 1
            continue
        }
            position_robo[0] += y
            position_robo[1] += x
            if (visited_robo[position_robo]) {
                visited_robo[position_robo] = visited_robo[position_robo] + 1
                continue
            }
            visited_robo[position_robo] = 1
    }
    const santa = Object.keys(visited_santa)
    const robo = Object.keys(visited_robo)

    const set = new Set([...santa, ...robo]).size
    console.log({ 
        set
    })

}

// part_one()
part_two()

File: 2015/03/main.exs
# Example values
#
# > = 2 -> initial house + 1
# ^>v< = 4 -> initial + 3 (initial receives two times)
# ^v^v^v^v^v = 2 -> initial + 1 (initial and above receives multiples times)
defmodule Resolve do
  def read_file(filename) do
    {_, file} = File.read(filename)
    file
      |> String.trim()
      |> String.split("")
      |> Enum.filter(fn letter -> String.length(letter) !== 0 end)
      |> Enum.map(fn letter -> 
        case letter do
          ">" -> [0, 1]
          "v" -> [-1, 0]
          "\^" -> [1, 0]
          "<" -> [0, -1]
        end
      end)
  end

  def part_one(file) do
    file
      |> Enum.scan(
        &([
          Enum.at(&1, 0) + Enum.at(&2, 0), 
          Enum.at(&1, 1) + Enum.at(&2, 1)
        ])
      )
      |> MapSet.new()
      |> MapSet.put([0,0])
      |> MapSet.size()
  end

  def part_two(file) do
    file
      |> Stream.with_index
      |> Enum.reduce([[],[]], fn ({x, i}, [evens, odds]) -> 
        case rem(i, 2) do
          0 -> [evens ++ [x], odds]
          _ -> [evens, odds ++ [x]]
        end
      end)
      |> Enum.map(fn point -> 
        point
        |> Enum.scan(
          &([
            Enum.at(&1, 0) + Enum.at(&2, 0), 
            Enum.at(&1, 1) + Enum.at(&2, 1)
          ])
        )
        |> MapSet.new()
        |> MapSet.to_list()
      end)
      |> Enum.flat_map(& &1)
      |> MapSet.new()
      |> MapSet.put([0,0])
      |> MapSet.size()
  end
end

file = Resolve.read_file("input")
values = Resolve.part_one(file)
splited_values = Resolve.part_two(file)

IO.inspect(values)
IO.inspect(splited_values, charlists: :as_lists)
